Accurate delayMicroseconds for odd frequencies

cburstedde opened the pull request:
Hi, just added delayMicroseconds for 18 MHz, which I'm currently using to overclock my tiny45.

What would be your opinion on the 18.432 MHz case, replacing the shifts
```c
    us = (us >> 1) + (us >> 2) + (us >> 3) + (us >> 4);
```
with the accurate long multiply
```c
    us = (us * 60398UL) >> 16; 
```
and adjusting the number of cycles used in calculating? (Not my idea, this is e.g. in MiniCore.)

cburstedde committed:
Add accurate delayMicroseconds for 18 MHz

SpenceKonde commented:
The multiply would run much faster on MiniCore - the ATmega parts have the AVRe+ core, with hardware multiplier, the ATtiny parts, with AVRe core, don't. I don't think it would have acceptable performance here.

cburstedde commented:
Makes sense, so leaving the 18.432 alone sounds fine. The 18 MHz case should be ready to go.

cburstedde committed:
Merge branch 'master' into delayMicroseconds18MHz

cburstedde committed:
Bring down delayMicroseconds error for 18.432 MHz

Upstream version computes (us * 15) >> 4 for an error of 17253 ppm.
Add alternative shift/add math that reduces error to 232 ppm (75x).

cburstedde committed:
Add accurate delayMicroseconds for 14.7456 Mhz

cburstedde committed:
Add accurate delayMicroseconds for 16.5 MHz

cburstedde commented:
Thinking about this some more, if delayMicroseconds busy waits anyway, we might as well use these instructions for accurate shift/add calculations. Turns out the changes were rather minor to give 18.432 MHz a 75x bump in accuracy. Its error dropped from 1 second per minute to below 1 second per hour.

While at it, I've added similar functionality to a couple other odd frequencies. All frequencies are still off by a few cycles, but with near zero drift for longish delays. In fact, 18 and 16.5 MHz now have exact zero drift.

I gave it enough testing to feel relatively final about it, but extra eyes won't hurt.

cburstedde committed:
Fix typo in comment

cburstedde committed:
Tweak to this PR's updates

cburstedde committed:
Tweak delayMicroseconds for 6 MHz and 4 Mhz

cburstedde committed:
Add accurate 2 MHz to delayMicroseconds()

SpenceKonde commented:
Looks safe enough, I'll take it. Thanks again!
